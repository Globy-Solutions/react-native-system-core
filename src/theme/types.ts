import type {Theme} from '@react-navigation/native/lib/typescript/src/types';
import type {Dispatch, ReactNode, SetStateAction} from 'react';
import type {ColorValue} from 'react-native/Libraries/StyleSheet/StyleSheet';
import type {TextStyle} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

export enum VariantFont {
  normal = 'normal',
  large = 'large',
  extraLarge = 'extraLarge',
  small = 'small',
  extraSmall = 'extraSmall'
}

export type ThemeProps = Theme & {
  dark: boolean;
  colors: {
    [key in keyof Theme['colors']]: ColorValue;
  };
  borderRadius: number;
  fonts: {
    [key in VariantFont]: TextStyle;
  };
  elevation?: number;
};

export type ThemeDataProps = ThemeProps & {
  isDark: boolean;
  changeTheme: Dispatch<SetStateAction<boolean>>;
};
export type ThemeProviderProps = {
  children: ReactNode;
};
