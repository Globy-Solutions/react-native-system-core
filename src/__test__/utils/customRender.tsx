import {render} from '@testing-library/react-native';
import {ErrorBoundary} from 'react-error-boundary';
import renderer from 'react-test-renderer';
import ThemeProvider from 'src/theme/theme-provider';

type Props = {
  children: React.ReactNode;
};

const logError = (error: Error, info: {componentStack: string}) => {
  // Do something with the error, e.g. log to an external API
  console.log(error, info);
};
const Provider = ({children}: Props) => (
  // @ts-ignore
  <ErrorBoundary FallbackComponent="Something went wrong" onError={logError}>
    <ThemeProvider>{children}</ThemeProvider>
  </ErrorBoundary>
);
const App = ({children}: Props) => (
  // @ts-ignore
  <ErrorBoundary FallbackComponent="Something went wrong" onError={logError}>
    {children}
  </ErrorBoundary>
);
const toJSON = (component: any) => {
  const r = App(component);
  renderer.create(r).toJSON();
  return render(r);
};

export * from '@testing-library/react-native';
export {App, Provider, toJSON};
