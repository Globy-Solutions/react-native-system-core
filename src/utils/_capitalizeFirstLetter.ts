export const capitalize = (text: string) => {
  const firstLetter = text.charAt(0);
  const firstLetterCap = firstLetter.toUpperCase();
  const remainingLetters = text.slice(1);
  return firstLetterCap + remainingLetters;
};

const capitalizeFirstLetter = (text: string) => {
  const textArr = /\b/i.test(text);
  if (textArr) {
    const arr = text.split(' ');
    return [arr[0], arr[1]]
      .map(t => capitalize(t as string).charAt(0))
      .join('');
  } else {
    return capitalize(text);
  }
};

export default capitalizeFirstLetter;
