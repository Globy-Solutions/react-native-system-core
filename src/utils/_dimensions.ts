import {Dimensions, PixelRatio} from 'react-native';

import type {DimensionValue} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

const {height, width}: {height?: DimensionValue; width?: DimensionValue} =
  Dimensions.get('screen');

const LINE_HEIGHT_SCALE_FACTOR = 1.33;

/**
 * Converts provided width percentage to independent pixel (dp).
 * @param  {string} widthPercent The percentage of screen's width that UI element should cover
 *                               along with the percentage symbol (%).
 * @return {number}              The calculated dp depending on current device's screen width.
 */
const widthPercentageToDP = (widthPercent: string | number) => {
  // Parse string percentage input and convert it to number.
  const elemWidth =
    typeof widthPercent === 'number' ? widthPercent : parseFloat(widthPercent);
  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((width * elemWidth) / 100);
};

/**
 * Converts provided height percentage to independent pixel (dp).
 * @param  {string} heightPercent The percentage of screen's height that UI element should cover
 *                                along with the percentage symbol (%).
 * @return {number}               The calculated dp depending on current device's screen height.
 */
const heightPercentageToDP = (heightPercent: string | number): number => {
  // Parse string percentage input and convert it to number.
  const elemHeight =
    typeof heightPercent === 'number'
      ? heightPercent
      : parseFloat(heightPercent);

  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((height * elemHeight) / 100);
};

/*
* Implement that piece of code to have orientation detection

Dimensions.get('screen').width < Dimensions.get('screen').height ? 'portrait' : 'landscape'
*/

const guidelineBaseWidth: number = 375;
const guidelineBaseHeight: number = 667;
const horizontalScale = (size: number) => (width / guidelineBaseWidth) * size;
const verticalScale = (size: number) => (height / guidelineBaseHeight) * size;
const moderateScale = (size: number, factor = 0.5) =>
  size + (horizontalScale(size) - size) * factor;

const getFontLineHeight = (fontSize: number) =>
  fontSize * LINE_HEIGHT_SCALE_FACTOR;

export {
  getFontLineHeight,
  height,
  horizontalScale,
  heightPercentageToDP as hp,
  moderateScale,
  verticalScale,
  width,
  widthPercentageToDP as wp
};
