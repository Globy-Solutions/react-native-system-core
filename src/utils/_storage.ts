/**
 * Method for LocalStorage
 * setData allways return Array object
 *
 * @function
 */

import {MMKV} from 'react-native-mmkv';

const AsyncStorage = new MMKV();

const getData = (name: string) => {
  const value = AsyncStorage.getString(name);
  return value ? JSON.parse(value) : null;
};

const setData = (name: string, data: any): void => {
  const currentState = getData(name);
  let temp: any[] = [];
  if (currentState) {
    temp = [...currentState, data];
  } else {
    temp = [data];
  }
  AsyncStorage.set(name, JSON.stringify(temp));
};

const deleteData = (name: string) => {
  AsyncStorage.delete(name);
};

export {deleteData, getData, setData};
