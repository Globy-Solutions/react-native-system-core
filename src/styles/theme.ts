import { Platform } from 'react-native';

import { getFontLineHeight, moderateScale } from '../utils/_dimensions';

import type { ThemeProps } from '../theme/types';

export const scale: number = Platform.OS === 'ios' ? 1 : 0.8;
export const fontFamily: string =
    Platform.OS === 'ios' ? 'Arial Hebrew' : 'Mooli Regular',
  fonts = {
    extraLarge: {
      fontFamily,
      fontSize: moderateScale(26 * scale),
      fontWeight: 'bold',
      lineHeight: getFontLineHeight(21 * scale),
      textTransform: 'capitalize'
    },
    extraSmall: {
      fontFamily,
      fontSize: moderateScale(12 * scale),
      lineHeight: getFontLineHeight(12 * scale)
    },
    large: {
      fontFamily,
      fontSize: moderateScale(22 * scale),
      fontWeight: 'bold',
      lineHeight: getFontLineHeight(20 * scale),
      textTransform: 'capitalize'
    },
    normal: {
      fontFamily,
      fontSize: moderateScale(18 * scale),
      lineHeight: getFontLineHeight(15 * scale)
    },
    small: {
      fontFamily,
      fontSize: moderateScale(16 * scale),
      lineHeight: getFontLineHeight(14 * scale)
    }
  } as const;
export const defaults = {
  borderRadius: 15,
  elevation: 5,
  fonts
};
export const lightTheme: ThemeProps = {
  colors: {
    primary: '#F3EF19',
    background: '#FFAF33',
    card: '#F38F07',
    text: '#115B99',
    border: '#074458',
    notification: '#FF5733'
  },
  dark: false,
  ...defaults
};

export const darkTheme: ThemeProps = {
  colors: {
    primary: '#28025B',
    background: '#09265D',
    card: '#5B3CF8',
    text: '#D7F2FB',
    border: '#fff',
    notification: '#EAAE07'
  },
  dark: true,
  ...defaults
};
