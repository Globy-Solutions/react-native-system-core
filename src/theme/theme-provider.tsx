/**
 * Renders a theme provider component that manages the theme state and provides it to its children.
 *
 * @param children - The child components to be wrapped by the theme provider.
 * @returns The rendered theme provider component.
 */

import {createContext, useContext, useEffect, useMemo, useState} from 'react';
import {Appearance} from 'react-native';

import {darkTheme, lightTheme} from '../styles/theme';

import type {ThemeDataProps, ThemeProps, ThemeProviderProps} from './types';

const ThemeProviderContext = createContext<ThemeDataProps | undefined>(
  undefined
);

export const useThemeProvider = (): ThemeDataProps => {
  const context = useContext(ThemeProviderContext);
  if (!context) {
    throw new Error('useThemeProvider must be used within a ThemeProvider');
  }
  return context;
};

export default function ThemeProvider({
  children
}: Readonly<ThemeProviderProps>): JSX.Element {
  const colorScheme = Appearance.getColorScheme();
  const [isDark, setIsDark] = useState<boolean>(false);
  const [theme, setTheme] = useState<ThemeProps>(lightTheme);

  useEffect(() => {
    setIsDark(colorScheme === 'dark');
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  useEffect(() => {
    isDark ? setTheme(darkTheme) : setTheme(lightTheme);
  }, [isDark]);

  const contextValue: ThemeDataProps = useMemo(
    () => ({
      changeTheme: setIsDark,
      isDark,
      ...theme
    }),
    [setIsDark, isDark, theme]
  );

  return (
    <ThemeProviderContext.Provider value={contextValue}>
      {children}
    </ThemeProviderContext.Provider>
  );
}
