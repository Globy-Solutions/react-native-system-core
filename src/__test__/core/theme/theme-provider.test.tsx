import '@testing-library/jest-native/extend-expect';
import React from 'react';
import {Appearance, Platform, Text, View} from 'react-native';

import {Provider, render, screen} from 'src/__test__/utils/customRender';
import {darkTheme, defaults, lightTheme} from 'src/styles/theme';
import {useThemeProvider} from 'src/theme/theme-provider';

import type {ReactTestInstance} from 'react-test-renderer';

const existOnScreen = (element: any) => {
  expect(element).toBeVisible();
  expect(element).toBeOnTheScreen();
  return element;
};
const Child = () => <Text>Test Child</Text>;
const Container = ({children}: {children: React.ReactNode}) => {
  const {
    colors: {background}
  } = useThemeProvider();
  return (
    <View style={{backgroundColor: background}} testID="container">
      {children}
    </View>
  );
};
const App = ({children}: {children?: React.ReactNode}) => (
  <Provider>{children}</Provider>
);

describe('fontFamily', () => {
  test('should return colors according to theme', () => {
    switch (Appearance.getColorScheme()) {
      case 'light':
        expect(lightTheme.dark).toBe(false);
        expect(lightTheme.colors.primary).toBe('#F3EF19');
        expect(lightTheme.colors.text).toBe('#115B99');
        break;
      case 'dark':
        expect(darkTheme.dark).toBe(true);
        expect(darkTheme.colors.primary).toBe('#28025B');
        expect(darkTheme.colors.text).toBe('#D7F2FB');
        break;
    }
  });

  test('should return the font family according to the OS', () => {
    const {
      fonts: {
        extraLarge: {fontFamily}
      }
    } = defaults;
    switch (Platform.OS) {
      case 'ios':
        expect(fontFamily).toBe('Arial Hebrew');
        break;
      case 'android':
        expect(fontFamily).toBe('Mooli Regular');
        break;
      default:
        expect(fontFamily).toBeUndefined();
    }
  });
});

describe('ThemeProvider', () => {
  describe('Child', () => {
    let text: ReactTestInstance;
    beforeEach(() => {
      render(<Child />);
      text = screen.getByText('Test Child');
    });
    test('snapshot', () => {
      expect(screen).toMatchSnapshot();
    });
    test('render content', () => {
      existOnScreen(text);
    });
  });
  describe('App', () => {
    let child: ReactTestInstance;
    let container: ReactTestInstance;
    beforeEach(() => {
      render(
        <App>
          <Container>
            <Child />
          </Container>
        </App>
      );
      child = screen.getByText('Test Child');
      container = screen.getByTestId('container');
    });
    test('snapshot', () => {
      expect(screen).toMatchSnapshot();
    });
    test('render content', () => {
      existOnScreen(screen.getByTestId('container'));
    });
    test('renders is visible', async () => {
      existOnScreen(child);
    });
    test('render styles', () => {
      const isDark = Appearance.getColorScheme() === 'dark';
      if (isDark) {
        expect(container).toHaveStyle({backgroundColor: '#09265D'});
        return;
      }
      expect(container).toHaveStyle({backgroundColor: '#FFAF33'});
    });
  });
});
