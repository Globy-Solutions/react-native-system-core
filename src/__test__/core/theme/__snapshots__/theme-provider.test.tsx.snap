// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider App snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFAF33",
    }
  }
  testID="container"
>
  <Text>
    Test Child
  </Text>
</View>
`;

exports[`ThemeProvider Child snapshot 1`] = `
<Text>
  Test Child
</Text>
`;
